### PR TITLE
x-pack/filebeat/input/httpjson: add transaction tracer

### LINKED ISSUE #32402: [Filebeat] httpjson - add option for capturing a trace data
**Describe the enhancement:**

As a user of the httpjson input that is developing configuration I would like to have a way of viewing the exact requests and responses associated with the input. When communicating in the clear with HTTP you can use tcpdump to see the raw requests and responses, but HTTPS is so common that this is rarely an option.

My proposal is to add an config option that writes timestamped requests and responses to a file via https://pkg.go.dev/net/http/httputil@go1.18.4#DumpRequestOut and https://pkg.go.dev/net/http/httputil@go1.18.4#DumpResponse. The file should automatically rotate based on size.

I considered adding an option to set a `KeyLogWriter` in the [tls.Config](https://pkg.go.dev/crypto/tls#Config) to allow decrypting a pcap of HTTPS traffic, but I think a plain text file will be user to use.

**Describe a specific use case for the enhancement or feature:**

When a user needs to debug the httpjson input during development they would enable the trace logging option to see the complete request and responses captured to a file.

### PR DESCRIPTION
**Pull request by efd6**
## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works — tested manually
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes #32402

### CODE REVIEW

File: x-pack/filebeat/input/httpjson/request.go [hunk 1]
```diff
@@ -57,6 +63,44 @@ func (c *httpClient) do(stdCtx context.Context, req *http.Request) (*http.Respon
...
+func (t *tracer) trace(req *http.Request, resp *http.Response) {
+	if t == nil {
+		return
+	}
+	t.txn++
+	b, err := httputil.DumpRequestOut(req, true)
+	if err != nil {
+		t.log.Print(err)
+	} else {
+		t.log.Printf("request %d.%d:\n\t%s", t.sess, t.txn, bytes.ReplaceAll(b, []byte{'\n'}, []byte("\n\t")))
```
Comment (belimawr) [inline x-pack/filebeat/input/httpjson/request.go:90]: [Question]
Why modify the original data? Why add this tab character at the beginning of new lines?

If the goal is to help debug issues, I believe it's better to keep the original data untouched.

File: x-pack/filebeat/input/httpjson/config_request.go [hunk 2]
```diff
@@ -12,6 +12,8 @@ import (
 	"strings"
 	"time"
 
+	"gopkg.in/natefinch/lumberjack.v2"
```
Comment (belimawr) [inline x-pack/filebeat/input/httpjson/config_request.go:15]: We already have a logging library in Beats: https://github.com/elastic/elastic-agent-libs/tree/main/logp that uses [go.uber.org/zap](https://pkg.go.dev/go.uber.org/zap) under the hood.

If possible I'd prefer to stick with those ones (preferably `elastic-agent-libs/logp`) to avoid introducing a new dependency and "logging" library.

What are the advantages of using lumberjack over the ones we already have in Beats?

Comment (andrewkroh) [reply x-pack/filebeat/input/httpjson/config_request.go:15]: Another alternative is https://pkg.go.dev/github.com/elastic/elastic-agent-libs/file#Rotator. It gives you an `io.WriteCloser` interface and manages rolling the underlying files.

But it looks like the library got polluted since I last looked with a hardcoded file extension 😢 . https://github.com/elastic/elastic-agent-libs/blob/a09d65fd12dec3efe53c44f05562431f96b1028b/file/rotator.go#L416  IMO is should have used `filepath.Ext` to get the extension from the configured `filename`.

File: x-pack/filebeat/input/httpjson/request.go (hunk 1 above)
Comment (efd6) [reply x-pack/filebeat/input/httpjson/request.go:90]: The log file is very hard to follow if it is not indented. This just moves the req and resp in one tab, essentially as the testing output does.

File: x-pack/filebeat/input/httpjson/config_request.go (hunk 2 above)
Comment (efd6) [reply x-pack/filebeat/input/httpjson/config_request.go:15]: > What are the advantages of using lumberjack over the ones we already have in Beats?

Simplicity; this does exactly what is needed and is transparently simple.

**Comment by efd6** [comment]
Failure is unrelated. It appears to have come in with https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/main/666/pipeline. Which does touch [code](https://github.com/elastic/beats/pull/32309/files#diff-55d635dc8516b02782b025d565e2191b6e5b559aedc21fd9437610aabaf739faR84-R99) that *could* explain it — it shows up first then — though I don't know enough about that part of libbeat to be confident of that and it passed the build at the time of the PR. @belimawr Do you think that could be the reason?

File: x-pack/filebeat/input/httpjson/request.go (hunk 1 above)
Comment (belimawr) [reply x-pack/filebeat/input/httpjson/request.go:90]: I get it now! Then it makes sense :smile:

File: x-pack/filebeat/input/httpjson/config_request.go (hunk 2 above)
Comment (belimawr) [reply x-pack/filebeat/input/httpjson/config_request.go:15]: @andrewkroh's suggestion looks great and simple. The file extension is a pitty, but we can easily fix that as well. I glanced over the code and it seems to be possible to make the extension configurable without changing the public API.

**Comment by belimawr** [comment]
> Failure is unrelated. It appears to have come in with https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/main/666/pipeline. Which does touch [code](https://github.com/elastic/beats/pull/32309/files#diff-55d635dc8516b02782b025d565e2191b6e5b559aedc21fd9437610aabaf739faR84-R99) that _could_ explain it — it shows up first then — though I don't know enough about that part of libbeat to be confident of that and it passed the build at the time of the PR. @belimawr Do you think that could be the reason?

I'll take a look at it right now.

**Comment by belimawr** [comment]
@efd6, yes, there is a bug on my bugfix :rofl: 

I'm writing the fix-fix PR right now. Another race condition. Thanks a lot for spotting that!!!

**Comment by belimawr** [comment]
> @efd6, yes, there is a bug on my bugfix rofl
> 
> I'm writing the fix-fix PR right now. Another race condition. Thanks a lot for spotting that!!!

Here is the fix PR: https://github.com/elastic/beats/pull/32433

**Review by andrewkroh — DISMISSED** [review]

File: x-pack/filebeat/input/httpjson/request.go [hunk 3]
```diff
@@ -42,6 +47,7 @@ func (c *httpClient) do(stdCtx context.Context, req *http.Request) (*http.Respon
 		return nil, fmt.Errorf("failed to execute http client.Do: %w", err)
 	}
 	defer resp.Body.Close()
+	c.tracer.trace(req, resp)
```
Comment (andrewkroh) [inline x-pack/filebeat/input/httpjson/request.go:50]: I think the request might need to be dumped before it is executed in order to access the body before it is drained.

Comment (efd6) [reply x-pack/filebeat/input/httpjson/request.go:50]: If you add this and run the tests you will see a log file with the request and response logging in "test.log".

```diff
diff --git a/x-pack/filebeat/input/httpjson/input_test.go b/x-pack/filebeat/input/httpjson/input_test.go
index 69617d82a5..d48f8d5ceb 100644
--- a/x-pack/filebeat/input/httpjson/input_test.go
+++ b/x-pack/filebeat/input/httpjson/input_test.go
@@ -583,6 +583,7 @@ func TestInput(t *testing.T) {
        for _, testCase := range testCases {
                tc := testCase
                t.Run(tc.name, func(t *testing.T) {
+                       tc.baseConfig["request.tracer.filename"] = "test.log"
                        tc.setupServer(t, tc.handler, tc.baseConfig)
 
                        cfg := conf.MustNewConfigFrom(tc.baseConfig)
```

**Review by andrewkroh — COMMENTED** [review]
I'm testing against a mock server that responds with an HTTP 503 in a particular case. When this happens the tracer logs "no response (`2022/07/25 11:28:46 no response 1658762511476.5`).

One idea I had was too see if there is a different hook point such that we could implement a http.RoundTripper for capturing the request/response. I think this could also give us visibility into requests made by retryablehttp... 🤔

**Comment by efd6** [comment]
That is possible though it adds a fair bit of complexity/indirection. I've fixed the loss of non-erroring non-200 status logging.

**Comment by andrewkroh** [comment]
I think the complexity is not too bad. I had to see for myself (it's just a hack), but have a look at https://github.com/elastic/beats/commit/93d5f2e84a5e67dbb9a297c6515f6b0c91e778c4  which builds upon your change and uses the `config.Request.Tracer`. The main benefit is that we can observe all requests and responses, which I think is really important when trying to understand what's happening during a debug session.

In my hack I did choose a structured logger. I figured if I get a big trace log from a user that I'm trying to help, then this format will be easier to analyze in Elasticsearch.

One feature that I thought would be nice to have is a `trace.id`. I set this up such that you can pass a `trace.id` to be logged in the http.Request.Context(). But I wasn't sure where is best to inject this. I think I would want to group the initial request, pagination requests, and chaining requests with the same trace.id. Then you can group all the data associated with one "iteration" of the input.

**Comment by efd6** [comment]
That's approximately the level of complexity that I was thinking it would involve when you include the code in go-examples. I agree that if the intention is for looking at large dumps then a structured logger would be better. Are you happy for me to pull the code you have in go-examples into beats?

One comment on the httplog, the logging of the body is not currently ECS compliant since it [dumps the complete serialised request](github.com/andrewkroh/go-examples), not the body (same for the response). To fix this we'd drop the use of httputil and just make a local drainBody to get those bytes. WDYT?

**Comment by andrewkroh** [comment]
Yeah, you are welcome to pull in any of that code. I did realize that I was including more than what is expected in those ECS `body.content` fields. I guess it would be easiest to create non-ECS fields like `http.{request,response}.Dump` (or some other name) than it would be reimplement parts of the httputil helper functions.

**Comment by efd6** [comment]
The helper is tiny so I'll make use of a [proverb](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s).

**Comment by efd6** [comment]
Initially in an internal package in httpjson, but I can imagine that this will have general utility, so it can be moved out later when needed elsewhere.

**Comment by efd6** [comment]
E2E test failure is unrelated.

File: x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+			zap.String("http.response.mime_type", resp.Header.Get("Content-Type")),
+		)
+	}
+	message, err = httputil.DumpResponse(resp, true)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to dump response: %w", err))
+	} else {
+		respParts = append(respParts, zap.ByteString("event.original", message))
+	}
+	if errors != nil {
```
Comment (andrewkroh) [inline x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go:153]: I think this needs to check for `len(errors) > 0`. In the case where the request produces an error and then we zero out the length this will be non-nil. In general I don't like to make a distinction between nil and empty slices so I tend to always check for length.

**Review by andrewkroh — COMMENTED** [review]
Thanks for making it ECS compliant.

File: x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+// nextTxID returns the next transaction.id value. It increments the internal
+// request counter.
+func (rt *LoggingRoundTripper) nextTxID() string {
+	count := rt.txIDCounter.Inc()
+	return rt.txBaseID + "-" + strconv.FormatUint(count, 10)
+}
+
+var noResponse = zap.NamedError("error.message", errors.New("unexpected nil response"))
```
[the remark this answers is not in the document]
Comment (efd6) [reply x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go:168]: I will check.

File: x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go (hunk 4 above)
Comment (efd6) [reply x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go:153]: Good point. I wrote the first one knowing that the zero length would be nil and then didn't reassess when I duplicated it.

File: x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (efd6) [reply x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go:168]: The two types of error messages:
```
{"log.level":"debug","@timestamp":"2022-07-28T21:08:54.725+0930","message":"HTTP request","transaction.id":"63GBEPIAVC2HE-1","url.original":"http://127.0.0.1:57587","url.scheme":"http","url.path":"","url.domain":"127.0.0.1","url.port":"57587","url.query":"","http.request.method":"POST","user_agent.original":"Elastic-Filebeat/8.4.0 (darwin; amd64; unknown; 0001-01-01 00:00:00 +0000 UTC)","error.message":[{"error":"failed to read request body: faux"},{"error":"failed to dump request: faux"}],"ecs.version":"1.6.0"}
```
and
```
{"log.level":"debug","@timestamp":"2022-07-28T21:12:09.042+0930","message":"HTTP response error","transaction.id":"R3VDV93NVC2HE-1","error":{"message":"unexpected nil response"},"ecs.version":"1.6.0"}
```
So the `noResponse` error looks like it is working as intended. I'll fix the other.

New version:
Single
```
{"log.level":"debug","@timestamp":"2022-07-28T21:24:05.124+0930","message":"HTTP request","transaction.id":"N34CJAOTVG2HE-3","url.original":"http://127.0.0.1:59402","url.scheme":"http","url.path":"","url.domain":"127.0.0.1","url.port":"59402","url.query":"","http.request.method":"GET","user_agent.original":"Elastic-Filebeat/8.4.0 (darwin; amd64; unknown; 0001-01-01 00:00:00 +0000 UTC)","event.original":"GET / HTTP/1.1\r\nHost: 127.0.0.1:59402\r\nUser-Agent: Elastic-Filebeat/8.4.0 (darwin; amd64; unknown; 0001-01-01 00:00:00 +0000 UTC)\r\nAccept: application/json\r\nAccept-Encoding: gzip\r\n\r\n","error.message":"failed to read request body: faux","ecs.version":"1.6.0"}
```
Double
```
{"log.level":"debug","@timestamp":"2022-07-28T21:24:02.121+0930","message":"HTTP request","transaction.id":"F31B1AOTVG2HE-2","url.original":"http://127.0.0.1:59398","url.scheme":"http","url.path":"","url.domain":"127.0.0.1","url.port":"59398","url.query":"","http.request.method":"GET","user_agent.original":"Elastic-Filebeat/8.4.0 (darwin; amd64; unknown; 0001-01-01 00:00:00 +0000 UTC)","error.message":["failed to read request body: faux","failed to dump request: faux"],"ecs.version":"1.6.0"}
```
Round trip
```
{"log.level":"debug","@timestamp":"2022-07-28T21:24:03.123+0930","message":"HTTP response error","transaction.id":"N34CJAOTVG2HE-2","error":{"message":"unexpected nil response"},"ecs.version":"1.6.0"}
```

[the remark this answers is not in the document]
Comment (andrewkroh) [reply x-pack/filebeat/input/httpjson/internal/httplog/roundtripper.go:176]: "New version" looks good.

**Review by andrewkroh — APPROVED** [review]
